types/patches/0022-adhoc-rnav-Manually-approximate-conditional-in-Route.patch

```diff
@@ -0,0 +1,55 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Greg Price <greg@zulip.com>
+Date: Tue, 24 May 2022 18:57:24 -0700
+Subject: [adhoc] rnav: Manually approximate conditional in Route
+
+In the upstream definition, this is a conditional type that's basically
+
+    undefined extends T ? { readonly x?: T } : { readonly x: T }
+
+That is, it's like `{ readonly x: T }`, but if T contains undefined,
+then the object could have `x` missing entirely as well as having
+the value `undefined`.
+
+I don't think there's a way to write precisely that type in Flow.
+Here, manually approximate it as `{ +x?: T }` -- that is, make the
+parameter always optional.
+
+Alternatively, we could write `{ +x: T }`.  For code consuming the
+value, I think that's nearly indistinguishable in Flow from the
+upstream type.  But then code supplying such a value would need an
+`x` property.  Here, for the `params` property on a route, that'd
+probably get annoying.
```

Comment:
Hmm, yeah, I guess Flow does have `$If` (as seen in the old code at the first of those two commits.)

... Or maybe it doesn't, and never did? https://flow.org/try just says "Cannot resolve name `$If`" if I try, and it's the same story if I try a bunch of older Flow versions there.

And I can't find any mention of it in the Flow source tree, even its history.

In fact, in the flow-typed project, its only appearance is in those react-navigation types.

So I suspect the story is that `$If` has never been a thing in Flow. And the author of those react-navigation libdefs was just fooled by the fact that until early 2021, Flow had a bug affecting libdefs where if a name couldn't be resolved it would suppress that error.

Reply:
Ha, weird.